syntax = "proto3";

package demo;

// 定义一个枚举类型
enum Sex {
  UNKNOWN = 0;
  MALE = 1;
  FEMALE = 2;
}

// 定义一个嵌套的消息
message UserInfo {
  string id = 1;
  string name = 2;
  int32 age = 3;
}

// 请求消息：包含多种数据类型
message HelloRequest {
  string greeting = 1; // 字符串
  int32 user_id = 2; // 整数
  bool is_active = 3; // 布尔值
  float score = 4; // 浮点数
  Sex sex = 5; // 枚举类型
  repeated string hobbies = 6; // 字符串列表
  map<string, string> properties = 7; // 键值对映射
  UserInfo user = 8; // 嵌套消息
}

// 响应消息：也包含多种数据类型
message HelloReply {
  string message = 1;
  int32 code = 2;
  repeated string log_messages = 3; // 列表
  UserInfo returned_user = 4; // 嵌套消息
}

service DemoService {
  rpc SayHello (HelloRequest) returns (HelloReply);
}
